atrium-api/src/types/string.rs

```diff
@@ -0,0 +1,285 @@
+/// around `String`.
+macro_rules! string_newtype {
+    ($name:ident) => {
+        impl FromStr for $name {
+            type Err = &'static str;
+
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                Self::new(s.into())
+            }
+        }
+
+        impl<'de> Deserialize<'de> for $name {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                let value = Deserialize::deserialize(deserializer)?;
+                Self::new(value).map_err(D::Error::custom)
+            }
+        }
+
+        impl Into<String> for $name {
+            fn into(self) -> String {
+                self.0
+            }
+        }
+
+        impl AsRef<str> for $name {
+            fn as_ref(&self) -> &str {
+                self.as_str()
+            }
+        }
+
+        impl Deref for $name {
+            type Target = str;
+
+            fn deref(&self) -> &Self::Target {
+                self.as_str()
+            }
+        }
+    };
+}
+
+/// An AT Protocol identifier.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum AtIdentifier {
+    Did(Did),
+    Handle(Handle),
+}
+
+impl From<Did> for AtIdentifier {
+    fn from(did: Did) -> Self {
+        AtIdentifier::Did(did)
+    }
+}
+
+impl From<Handle> for AtIdentifier {
+    fn from(handle: Handle) -> Self {
+        AtIdentifier::Handle(handle)
+    }
+}
+
+impl FromStr for AtIdentifier {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(did) = s.parse() {
+            Ok(AtIdentifier::Did(did))
+        } else {
+            s.parse().map(AtIdentifier::Handle)
+        }
+    }
+}
+
+impl Into<String> for AtIdentifier {
+    fn into(self) -> String {
+        match self {
+            AtIdentifier::Did(did) => did.into(),
+            AtIdentifier::Handle(handle) => handle.into(),
+        }
+    }
+}
+
+impl AsRef<str> for AtIdentifier {
+    fn as_ref(&self) -> &str {
+        match self {
+            AtIdentifier::Did(did) => did.as_ref(),
+            AtIdentifier::Handle(handle) => handle.as_ref(),
+        }
+    }
+}
+
+/// A generic [DID Identifier].
+///
+/// [DID Identifier]: https://atproto.com/specs/did
+#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
+#[serde(transparent)]
+pub struct Did(String);
+string_newtype!(Did);
+
+impl Did {
+    /// Parses a `Did` from the given string.
+    pub fn new(did: String) -> Result<Self, &'static str> {
+        const RE_DID: OnceCell<Regex> = OnceCell::new();
+
+        // https://atproto.com/specs/did#at-protocol-did-identifier-syntax
+        if !RE_DID
+            .get_or_init(|| Regex::new(r"^did:[a-z]+:[a-zA-Z0-9._:%-]*[a-zA-Z0-9._-]$").unwrap())
+            .is_match(&did)
+        {
+            Err("Invalid DID")
+        } else if did.len() > 2048 {
```

Comment:
Oh yeah, indeed it would be. I think I just copy-pasted this from a different parser where the initial parse was necesasry to get the string that would have its length be measured. But indeed here the length can be checked first.